File: math/tests/quantity.py
""" Checks atom methods and attributes. """
def test():
  """ Test atom initialization. """
  from _quantity import is_quantity, fromC, fromPy, get_angstrom, as_real,     \
                        get_as
  from numpy import abs
  from quantities import meter, angstrom, eV, nanometer# , Ry

  assert not is_quantity("a");
  assert not is_quantity(0);
  assert not is_quantity(0.4);
  assert is_quantity(eV)
  assert is_quantity(5*eV)
  assert is_quantity(5.45*nanometer)

  assert abs(fromC(5, 'meter') - 5*meter) < 1e-8
  assert abs(fromC(float((5*meter).rescale(angstrom)), 'angstrom') - 5*meter) < 1e-8
  assert abs(fromC(float((4*meter).rescale(angstrom)), 'angstrom') - 5*meter) > 1e-8
  assert abs(fromPy(5, meter) - 5*meter) < 1e-8
  assert hasattr(fromPy(5, meter), 'rescale')
  assert abs(get_angstrom(5*meter) - float(5*meter.rescale(angstrom))) < 1e-8
  assert abs(get_angstrom(5) - 5) < 1e-8
  assert abs(get_angstrom(5.5) - 5.5) < 1e-8
  assert get_angstrom(4*eV) is None
  assert abs(as_real(4.5*eV) - 4.5) < 1e-8
  assert as_real('a') is None

  assert abs(get_as(5*meter, angstrom) - float(5*meter.rescale(angstrom))) < 1e-8
  assert get_as(5*meter, eV) is None
  assert get_as(5*meter, 'a') is None
  assert get_as('5*meter', angstrom) is None
  assert abs(get_as(5*meter, 4*angstrom) - 5e10) < 1e-4

if __name__ == "__main__":
  from sys import argv, path 
  if len(argv) > 0: path.extend(argv[1:])
  test()

File: math/tests/quantity.cc
#include "LaDaConfig.h"

#include "../quantity.h"


using namespace LaDa::math;
PyObject* is_quantity(PyObject *_module, PyObject *_in)
{ 
  if(not PyQuantity_Check(_in)) Py_RETURN_FALSE;
  Py_RETURN_TRUE;
}
PyObject* fromC(PyObject *_module, PyObject *_args)
{
  double pynumber;
  char *pyunits;
  if(not PyArg_ParseTuple(_args, "ds", &pynumber, &pyunits)) return NULL;
  return PyQuantity_FromC(pynumber, std::string(pyunits));
}
PyObject* fromPy(PyObject *_module, PyObject *_args)
{
  PyObject *number;
  PyObject *units;
  if(not PyArg_ParseTuple(_args, "OO", &number, &units)) return NULL;
  return PyQuantity_FromPy(number, units);
}
PyObject* get_angstrom(PyObject *_module, PyObject *_in)
{
  LaDa::types::t_real result(PyQuantity_Get(_in, "angstrom"));
  if(std::abs(result) < 1e-8 and PyErr_Occurred())
  {
    PyErr_Clear();
    Py_RETURN_NONE;
  }
  return PyFloat_FromDouble(result);
}

#ifndef PyMODINIT_FUNC	/* declarations for DLL import/export */
# define PyMODINIT_FUNC void
#endif

#ifdef LADA_DECLARE
#  error LADA_DECLARE already defined.
#endif
#define LADA_DECLARE(name, args) {#name, (PyCFunction)name, METH_ ## args, ""} 

static PyMethodDef methods[] = { 
  LADA_DECLARE(is_quantity, O),
  LADA_DECLARE(fromC, VARARGS),
  LADA_DECLARE(fromPy, VARARGS),
  LADA_DECLARE(get_angstrom, O),
  {NULL},
};

#undef LADA_DECLARE

PyMODINIT_FUNC init_quantity(void) 
{
  PyObject* module = Py_InitModule("_quantity", methods);
}
